BaseClasses.py

```diff
@@ -760,6 +760,14 @@ def has_any(self, items: Iterable[str], player: int) -> bool:
         """Returns True if at least one item name of items is in state at least once."""
         return any(self.prog_items[player][item] for item in items)
 
+    def has_all_counts(self, item_counts: Dict[str, int], player: int) -> bool:
```

Comment:
Usually, if you don't need a parameter to be mutable, it's better to use the immutable interface (`Mapping` instead of `Dict`).

`item_counts` is not mutated in this function.

With the mutable type annotation, if someone has a more specific type (like a `TypedDict`), then the type checker points out that it's not safe to pass that more specific type to this function, because this function might mutate it and invalidate the typing.